pkg/relayer/proxy/websockets/bridge.go

```diff
@@ -0,0 +1,379 @@
+			if block.Height() >= closeHeight {
+				b.logger.Info().Msg("session closing, bridge stopped")
+				b.cancelCtx()
+			}
+		},
+	)
+
+	b.logger.Info().Msg("bridge started")
+}
+
+// messageLoop is the main loop of the bridge:
+//   - It listens for messages from the gateway and forwards them to the service backend,
+//   - It listens for messages from the service backend and forwards them to the gateway.
+func (b *bridge) messageLoop() {
+	for {
+		select {
+		case <-b.ctx.Done():
+			return
+		case msg := <-b.msgChan:
+			switch msg.source {
+
+			case messageSourceGateway:
+				b.handleGatewayIncomingMessage(msg)
+
+			case messageSourceServiceBackend:
+				b.handleServiceBackendIncomingMessage(msg)
+			}
+		}
+	}
+}
+
+// handleGatewayIncomingMessage handles incoming messages from the gateway.
+// It receives relay requests from the gateway, verifies them, forwards their payloads
+// to the service backend.
+func (b *bridge) handleGatewayIncomingMessage(msg message) {
+	logger := b.logger.With(
+		"message_source", messageSourceGateway,
+		"message_type", msg.messageType,
+	)
+
+	logger.Debug().Msg("received message from gateway")
+
+	// Unmarshal msg.data into a RelayRequest.
+	var relayRequest types.RelayRequest
+	if err := relayRequest.Unmarshal(msg.data); err != nil {
+		b.serviceBackendConn.handleError(
+			ErrWebsocketsGatewayMessage.Wrapf("failed to unmarshal relay request: %v", err),
+		)
+		return
+	}
+	serviceId := relayRequest.Meta.SessionHeader.ServiceId
+
+	// Store the latest relay request to guarantee that there is always a request/response.
+	b.setLatestRelayRequest(&relayRequest)
+
+	relayer.RelaysTotal.With(
+		"service_id", serviceId,
+		"supplier_operator_address", relayRequest.Meta.SupplierOperatorAddress,
+	).Add(1)
+
+	relayer.RelayRequestSizeBytes.With("service_id", serviceId).
+		Observe(float64(relayRequest.Size()))
+
+	// Verify the relay request signature and session.
+	if err := b.relayAuthenticator.VerifyRelayRequest(b.ctx, &relayRequest, serviceId); err != nil {
+		b.serviceBackendConn.handleError(
+			ErrWebsocketsGatewayMessage.Wrapf("failed to verify relay request: %v", err),
+		)
+		return
+	}
+
+	logger.Debug().Msg("relay request verified")
+
+	// Forward the relay request payload to the service backend.
+	if err := b.serviceBackendConn.WriteMessage(msg.messageType, relayRequest.Payload); err != nil {
+		b.serviceBackendConn.handleError(
+			ErrWebsocketsGatewayMessage.Wrapf("failed to send relay request to service backend: %v", err),
+		)
+		return
+	}
+
+	logger.Debug().Msg("relay request forwarded to service backend")
+
+	// Do not emit a relay to the miner if there is no response to form a request/response pair.
+	if b.getLatestRelayResponse() == nil {
+		logger.Info().Msg("waiting for service backend response")
+		return
+	}
+
+	relay := &types.Relay{
+		Req: &relayRequest,
+		Res: b.getLatestRelayResponse(),
+	}
+
+	relayer.RelaysSuccessTotal.With("service_id", serviceId).Add(1)
```

Comment:
Are these all already visible and tracklable in grafana?

If so - can you add a screenshot?
If not - can you add a TODO_MAINNET?